models/DataObject/ClassDefinition/Data/Extension/Relation.php

```diff
@@ -26,49 +35,88 @@ trait Relation
-            } elseif (is_array($documentTypes)) {
+            if ($documentTypes = $this->getDocumentTypes()) {
                 foreach ($documentTypes as $item) {
-                    $class[] = sprintf('\Pimcore\Model\Document\%s', ucfirst($item['documentTypes']) . $strArray);
+                    $types[] = $this->getClassName('document', $item['documentTypes']);
                 }
+            } else {
+                $types[] = Page::class;
+                $types[] = Snippet::class;
+                $types[] = Document::class;
             }
         }
 
-        // add asset
+        // add assets
         if ($this->getAssetsAllowed()) {
-            $assetTypes = $this->getAssetTypes();
-            if (count($assetTypes) == 0) {
-                $class[] = '\Pimcore\Model\Asset' . $strArray;
-            } elseif (is_array($assetTypes)) {
+            if ($assetTypes = $this->getAssetTypes()) {
                 foreach ($assetTypes as $item) {
-                    $class[] = sprintf('\Pimcore\Model\Asset\%s', ucfirst($item['assetTypes']) . $strArray);
+                    $types[] = $this->getClassName('asset', $item['assetTypes']);
                 }
+            } else {
+                $types[] = Asset::class;
             }
         }
 
         // add objects
         if ($this->getObjectsAllowed()) {
-            $classes = $this->getClasses();
-            if (count($classes) === 0) {
-                $class[] = '\Pimcore\Model\DataObject\AbstractObject' . $strArray;
-            } elseif (is_array($classes)) {
+            if ($classes = $this->getClasses()) {
                 foreach ($classes as $item) {
-                    $class[] = sprintf('\Pimcore\Model\DataObject\%s', ucfirst($item['classes']) . $strArray);
+                    $types[] = $this->getClassName('object', $item['classes']);
                 }
+            } else {
+                $types[] = AbstractObject::class;
             }
         }
 
-        return $class;
+        $types = array_map(static fn (string $type): string => u($type)->ensureStart('\\')->toString(), $types);
+
+        if ($asArray) {
+            $types = array_map(static fn (string $type): string => $type . '[]', $types);
+        }
+
+        return $types;
+    }
+
+    /**
+     * @param 'asset'|'document'|'object' $type
+     */
+    private function getClassName(string $type, string $shortName): string
+    {
+        $typeLoader = match ($type) {
+            'document' => \Pimcore::getContainer()->get(DocumentTypeLoader::class),
+            default => null,
+        };
+
+        if ($typeLoader) {
+            try {
+                return $typeLoader->getClassNameFor($shortName);
+            } catch (UnsupportedException) {
+                // try next
+            }
+
+            try {
+                return $typeLoader->build($shortName)::class;
+            } catch (UnsupportedException) {
+                // try next
+            }
```

Comment:
I'm not entirely sure if we should do this. What happens if the constructor has mandatory arguments? Could that happen?